### ISSUE TITLE
Read Headers on the received frame

### ISSUE THREAD

**Issue by akashdk191** [tarasko/picows#22]
Hii Tarasko.
 I was checking the example program. And I am planning to connect my EV charger simulator running on OCPP1.6 protocol over websockets. Hence I must read the incoming protocol format, for which I must read the header of the WSTransport, sSomething like this websocket.request_headers['Sec-WebSocket-Protocol']. Would like to know how to read the headers in picows?

**Comment by tarasko** [comment]
Currently, picows doesn't support protocols, because I personally didn't need them, but I think it would be is easy to add.

Sorry, I'm not familiar with OCPP protocol, is your use case websocket client or websocket server?

**Comment by akashdk191** [comment]
I tried few modifications on library but couldn't succeed. My case is a  Websocket based OCPP server. OCPP1.6 protocol is nothing complicated, just JSON request and response. 
But Websocket must have support to 
1. Read Header from received frame (to determine protocol)
2. Certificate based encryption
Can this be done using the picows. It would be very helpful in case you can add interface to read headers

**Comment by tarasko** [comment]
1. Just want to mention this so that we are on the same page.
HTTP header 'Sec-WebSocket-Protocol' is only send only once by the client at the negotiation phase. The server shall reply with its own 'Sec-WebSocket-Protocol' with the list of supported protocols. Websocket frames are sent without headers.

When implementing a server you can check the content of 'Sec-WebSocket-Protocol'. WSUpgradeRequest object with headers is passed to ws_protocol_factory function.

```
    class ServerClientListener(WSListener):
        ...

    def listener_factory(request: WSUpgradeRequest):
        print(request.headers["Sec-WebSocket-Protocol"]
        return ServerClientListener()

    plain_server = await ws_create_server(listener_factory,
                                          "127.0.0.1", 9001,
                                          websocket_handshake_timeout=0.5)
```
The problem is that the server is always replying with empty "Sec-WebSocket-Protocol" to the client request. This is something that has to be changed. You should be able to specify what protocols are supported by the server from listener_factory.

2. By "certificate based encryption" do you mean standard SSL/TLS? 

If so, then yes it supported. 
You would have to create SSLContext object and pass it to ws_create_server.

https://github.com/tarasko/picows/blob/master/tests/test_basics.py#L60

**Comment by akashdk191** [comment]
> 1. HTTP header 'Sec-WebSocket-Protocol' is sent only once by the client at the negotiation phase.
> 
> The server shall reply with its own 'Sec-WebSocket-Protocol' with the list of supported protocols. Websocket frames are sent without headers.
> 
> When implementing a server you can check the content of 'Sec-WebSocket-Protocol'. WSUpgradeRequest object with headers is passed to ws_protocol_factory function.
> 
> ```
>     class ServerClientListener(WSListener):
>         ...
> 
>     def listener_factory(request: WSUpgradeRequest):
>         print(request.headers["Sec-WebSocket-Protocol"]
>         return ServerClientListener()
> 
>     plain_server = await ws_create_server(listener_factory,
>                                           "127.0.0.1", 9001,
>                                           websocket_handshake_timeout=0.5)
> ```
> 
> The problem is that the server is always replying with empty "Sec-WebSocket-Protocol" to the client request. This is something that has to be changed. You should be able to specify what protocols are supported by the server from listener_factory.
> 
> 2. By "certificate based encryption" do you mean standard SSL/TLS?
> 
> If so, then yes it is supported. You would have to create SSLContext object and pass it to ws_create_server.
> 
> https://github.com/tarasko/picows/blob/master/tests/test_basics.py#L60

**Comment by akashdk191** [comment]
> > 1. HTTP header 'Sec-WebSocket-Protocol' is sent only once by the client at the negotiation phase.
> > 
> > The server shall reply with its own 'Sec-WebSocket-Protocol' with the list of supported protocols. Websocket frames are sent without headers.
> > When implementing a server you can check the content of 'Sec-WebSocket-Protocol'. WSUpgradeRequest object with headers is passed to ws_protocol_factory function.
> > ```
> >     class ServerClientListener(WSListener):
> >         ...
> > 
> >     def listener_factory(request: WSUpgradeRequest):
> >         print(request.headers["Sec-WebSocket-Protocol"]
> >         return ServerClientListener()
> > 
> >     plain_server = await ws_create_server(listener_factory,
> >                                           "127.0.0.1", 9001,
> >                                           websocket_handshake_timeout=0.5)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The problem is that the server is always replying with empty "Sec-WebSocket-Protocol" to the client request. This is something that has to be changed. You should be able to specify what protocols are supported by the server from listener_factory.
> > 
> > 2. By "certificate based encryption" do you mean standard SSL/TLS?
> > 
> > If so, then yes it is supported. You would have to create SSLContext object and pass it to ws_create_server.
> > https://github.com/tarasko/picows/blob/master/tests/test_basics.py#L60


Thanks for the response. Sure I will check the header during the negotiation phase. hopefully I will able to integrate with OCPP server. 
Yes the standard TLS support

**Comment by tarasko** [comment]
Hi @akashdk191, I've just released 1.7.1 where you negotiate protocol both with server and client.
Checkout an example https://github.com/tarasko/picows/blob/master/examples/subprotocol_negotiation.py
